File: backend/src/main/java/com/example/demo/service/CouponService.java
package com.example.demo.service;

import com.example.demo.entity.Coupon;
import com.example.demo.repository.CouponRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
@RequiredArgsConstructor
public class CouponService {
    private final CouponRepository couponRepository;

    public List<Coupon> getAllCoupons() {
        return couponRepository.findAll();
    }

    public Coupon createCoupon(Coupon coupon) {
        return couponRepository.save(coupon);
    }

    public Coupon getCouponById(Integer id) {
        return couponRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Coupon not found"));
    }

    public Coupon applyCoupon(String code) {
        Coupon coupon = couponRepository.findByCode(code)
                .orElseThrow(() -> new RuntimeException("Coupon not found"));

        if (!coupon.getIsActive() || coupon.getExpiryDate().isBefore(LocalDate.now())) {
            throw new RuntimeException("Coupon is invalid or expired");
        }

        return coupon;
    }

    public Coupon updateCoupon(Integer id, Coupon couponDetails) {
        Coupon existing = getCouponById(id);
        existing.setCode(couponDetails.getCode());
        existing.setDiscountPercent(couponDetails.getDiscountPercent());
        existing.setExpiryDate(couponDetails.getExpiryDate());
        existing.setIsActive(couponDetails.getIsActive());
        return couponRepository.save(existing);
    }

    public void deleteCoupon(Integer id) {
        couponRepository.deleteById(id);
    }
}

File: frontend/src/services/coupon.service.js
import api from './api';

export const couponService = {
  // Lấy danh sách mã giảm giá
  getAllCoupons: () => {
    return api.get('/coupons');
  },

  // Thêm mã giảm giá mới
  createCoupon: (couponData) => {
    return api.post('/coupons', couponData);
  },

  // Áp dụng mã giảm giá (Dành cho user)
  applyCoupon: (code) => {
    return api.get(`/coupons/apply/${code}`);
  },

  // Cập nhật mã giảm giá
  updateCoupon: (id, couponData) => {
    return api.put(`/coupons/${id}`, couponData);
  },

  // Xóa mã giảm giá
  deleteCoupon: (id) => {
    return api.delete(`/coupons/${id}`);
  }
};

